Fix Kamino transactions endpoint

peroxy opened the pull request:
1. Fetch ATAs from strategy state instead of transaction accounts (those are out of order because of IDL changes)
2. Add total SOL from all SOL atas to fix the issue where SOL ata had 0 tokens

peroxy commented on src/constants/redis.ts:
```diff
@@ -5,6 +5,7 @@ export const BORROWING_VERSION_EXPIRY_IN_SECONDS = 60;
 export const LOANS_EXPIRY_IN_SECONDS = 60;
 export const METRICS_EXPIRY_IN_SECONDS = 5 * 60;
 export const STRATEGY_HISTORY_EXPIRY_IN_SECONDS = 5 * 60;
+export const STRATEGIES_EXPIRY_IN_SECONDS = 60 * 60 * 24;
```
Strategy state properties that we use (token mints) shouldn't change at all, that's why I cache them for 24 hours

y2kappa commented on src/api/kamino/transactions.ts:
```diff
@@ -162,18 +212,10 @@ async function getTokenInfo(timestamp: Date, tokenA: any, config: HubbleConfig,
+function getTokenAmount(token: any) {
+  return token
+    ? Decimal.abs(new Decimal(token.pre).minus(token.post)).dividedBy(Decimal.pow(10, token.decimals))
     : new Decimal(0);
```
Name of the function is too generic to mean what it actually does.

Also, I would throw or return a null that you can handle, returning 0 which could be a valid number is very difficult to track and late to discover.

peroxy replied:
Renamed the function and returned null now - these errors get logged and the transaction skipped

y2kappa commented on src/api/kamino/transactions.ts:
```diff
@@ -68,55 +69,98 @@ async function fetchTransactions(
+    let tokenBAmount = getTokenAmount(tokenB);
+
+    if (tokenAIsSol) {
+      tokenAAmount = new Decimal(0);
```
I would not reset it to 0, I would just add on top of it. This logic doesn't work in all cases.

y2kappa replied:
I assume this transaction will fail: https://explorer.solana.com/tx/2iXseS1QpK58UJbHnmfPrWxZmcXjvgQV1FpVbRqo1b3mPEG6QbZBaqWJUwLuG2ct3tnafHLLC8ThdRKHg5kRKyRA

What does this new code output for these txns:

https://solana.fm/tx/hKPq3VUynLYgcTU45odaRx3Cno67ZF6qkPyqoKgd9wduXU6MVzzQfkvYKwzJMZog7vGXcH8jhe33qzbeqVXXGUw
https://solana.fm/tx/4HrGWQb6Ek1saqEqu7Lo5vtQU7FNzkdkacskESuiM9uUzqXmjLBiNJQJyvi4uJCK4ZcBHM3aPhW7sViGN2JRq8xz
https://solana.fm/tx/Ed4YESigpxLouqXVbmGCfw7T7HAQHNC9dNoVqecccfu6JTtq8bmB4GfqghnEG3XQu988FVGKkqDreW4Y6ESwPzd
https://solana.fm/tx/3DzJuviUor15PYZovKJuD2gajysX34P54FCCzXe692ULUiDntZwALFzwA4owgJkhJqhGooA1DThE93EpN1edrZmC
https://solana.fm/tx/63t5U6EozwVS5J1EcYZ5xavNMpttojpQupkscQPFmcr53BggkBNVc1JVmF5czmRjcpR4UdvLdyFos2dDpjkqDvkF

Keep in mind that not in all cases if it's sol or wsol it fails to parse. In some cases it takes correctly the diff from the wsol ata

<img width="832" alt="Screenshot 2023-01-24 at 08 15 33" src="https://user-images.githubusercontent.com/10101287/214242144-e05d00fb-13f7-469a-8a8e-c084b4c49d9c.png">

y2kappa commented on src/api/kamino/transactions.ts:
```diff
@@ -68,55 +69,119 @@ async function fetchTransactions(
+  shareholder: PublicKey
 ): Promise<KaminoTransactionResponse[]> {
   const response: KaminoTransactionResponse[] = [];
+  const solMint = config.kamino.mints.find((x) => x.scopeToken === 'SOL')!.address;
```
This should be a const, Scope is a very unnecessary dependency here. There are two mints related to sol. Native mint and wrapped sol mint. That's all. We only use the wrapped sol mint and that's a const across solana and all programs.

y2kappa replied:
```suggestion
  const wsolMint = NATIVE_MINT; // if i remember correctly
```

y2kappa commented on src/api/kamino/transactions.ts:
```diff
@@ -68,55 +69,119 @@ async function fetchTransactions(
+      }
+    );
+    if (!strategy) {
+      logger.error(`Could not find strategy ${strategyPubkey.pubKey}`);
```
no silent errors, this should throw, this is fatal

y2kappa commented on src/api/kamino/transactions.ts:
```diff
@@ -68,55 +69,119 @@ async function fetchTransactions(
+    const numberOfShares = getTokenDifferenceFromPrePostBalance(shares);
+    if (numberOfShares === null) {
+      logger.error(`Could not get token difference from token pre/post balances for number of shares`);
+      continue;
```
throw or return error, no silent errors here please

y2kappa commented on src/api/kamino/transactions.ts:
```diff
@@ -68,55 +69,119 @@ async function fetchTransactions(
+          }
+        }
+      } else {
+        logger.error(`Could not get token difference from token pre/post balances for token A`);
```
log what is tokenA, the strategy, the params of the request, for debugging purposes.
in the else statement -> throw, this case should never happen, if it happens we crash

y2kappa replied:
the person that needs to handle this is the app, not the api. the app should try catch around a throwing api

y2kappa commented on src/api/kamino/transactions.ts:
```diff
@@ -68,55 +69,119 @@ async function fetchTransactions(
+          }
+        }
+      } else {
+        logger.error(`Could not get token difference from token pre/post balances for token B`);
```
same comments as above

y2kappa commented on src/api/kamino/transactions.ts:
```diff
@@ -68,55 +69,119 @@ async function fetchTransactions(
-    const { numberOfShares, tokenAAmount, tokenBAmount } = getTokenAmounts(shares, tokenA, tokenB);
+    const userSharesAta = await getAssociatedTokenAddress(new PublicKey(strategy.sharesMint), shareholder);
+    const tokenAAta = await getAssociatedTokenAddress(new PublicKey(strategy.tokenAMint), shareholder);
+    const tokenBAta = await getAssociatedTokenAddress(new PublicKey(strategy.tokenBMint), shareholder);
```
also create wsolAta here and check below if necessary

y2kappa commented on src/api/kamino/transactions.ts:
```diff
@@ -68,55 +69,119 @@ async function fetchTransactions(
+      if (tokenAIsSol) {
+        tokenAAmount = new Decimal(0);
+        for (const ata of Object.keys(instruction.txBalances)) {
+          if (instruction.txBalances[ata]?.token?.mint === solMint.toString()) {
```
&& address is wsolAta

y2kappa replied:
```suggestion
          if (instruction.txBalances[ata]?.token?.mint === wsolMint.toString()) {
```

y2kappa replied:
I don't think this is right, we had a txn in which there was no `mint` yet the user was paying SOL from his personal wallet.

y2kappa commented on src/api/kamino/transactions.ts:
```diff
@@ -68,55 +69,119 @@ async function fetchTransactions(
+  shareholder: PublicKey
 ): Promise<KaminoTransactionResponse[]> {
   const response: KaminoTransactionResponse[] = [];
+  const solMint = config.kamino.mints.find((x) => x.scopeToken === 'SOL')!.address;
```
Logic should be:

let solDiff = ...
let wsolDiff = ...
let tokenAAmount = if exists tokenADiff else if tokenA = sol then solDiff + wsolDiff else throw;
let tokenBAmount = if exists tokenBDiff else if tokenB = sol then solDiff + wsolDiff else throw;

peroxy commented:
@y2kappa addressed all your comments, here's what this response looks like:

```shell
curl --location --request GET 'localhost:8888/shareholders/79nBgvQX8EdGxJbWmnVArrihTCZwSdsqLoyUzUHRJxuw/transactions?env=mainnet-beta'
```

response:
```json
{
    "transactions": [
        {
            "timestamp": "2023-01-11T18:23:38.788Z",
            "transactionSignature": "2Hzsgt8ChPNC8HMzMR45mGnz3C9TGqkSa4oSCNf8KsvsCMds1fcEWeM7a1j6feh8VhEJbLiKMvGyDrEYEbSizAHB",
            "transactionName": "depositAndInvest",
            "strategy": "HWg7yB3C1BnmTKFMU3KGD7E96xx2rUhv4gxrwbZLXHBt",
            "tokenA": "SOL",
            "tokenAAmount": "0.023815696",
            "tokenAPrice": "15.7147",
            "tokenB": "BONK",
            "tokenBAmount": "471270.83303",
            "tokenBPrice": "0.0000015434",
            "usdValue": "1.101615921629702",
            "numberOfShares": "1.105027"
        },
        {
            "timestamp": "2023-01-11T18:20:11.612Z",
            "transactionSignature": "4FMNLSQzv6zs2Vt9ee4X2YWz4GuSi6PaHnDNR31FrhMLLNNi41qi9gER9MCpcTxSMUTj9JQQLNvLF7XMXTxKJtAj",
            "transactionName": "depositAndInvest",
            "strategy": "ByXB4xCxVhmUEmQj3Ut7byZ1Hbva1zhKjaVcv3jBMN7E",
            "tokenA": "SOL",
            "tokenAAmount": "0.442798657",
            "tokenAPrice": "15.715",
            "tokenB": "STSOL",
            "tokenBAmount": "0.051688286",
            "tokenBPrice": "16.9847927257",
            "usdValue": "7.8364957188117011502",
            "numberOfShares": "0.021301113"
        },
        {
            "timestamp": "2023-01-11T17:36:20.835Z",
            "transactionSignature": "63t5U6EozwVS5J1EcYZ5xavNMpttojpQupkscQPFmcr53BggkBNVc1JVmF5czmRjcpR4UdvLdyFos2dDpjkqDvkF",
            "transactionName": "depositAndInvest",
            "strategy": "ByXB4xCxVhmUEmQj3Ut7byZ1Hbva1zhKjaVcv3jBMN7E",
            "tokenA": "SOL",
            "tokenAAmount": "0.179292172",
            "tokenAPrice": "15.48246305",
            "tokenB": "STSOL",
            "tokenBAmount": "0.019149818",
            "tokenBPrice": "16.7258482826",
            "usdValue": "3.0961813786516471668",
            "numberOfShares": "0.008542911"
        },
        {
            "timestamp": "2023-01-06T21:14:05.835Z",
            "transactionSignature": "4nusLE1XCE8ZcrFzeMsr2ZjDLzUJrcPPq9bBUxdR1TMnnZytuCP2XhBwsUwd7JJdJtKpbRQMmuPjmgat2qyfRVR1",
            "transactionName": "depositAndInvest",
            "strategy": "HWg7yB3C1BnmTKFMU3KGD7E96xx2rUhv4gxrwbZLXHBt",
            "tokenA": "SOL",
            "tokenAAmount": "2",
            "tokenAPrice": "13.23409134",
            "tokenB": "BONK",
            "tokenBAmount": "23190470.76525",
            "tokenBPrice": "0.0000020638",
            "usdValue": "74.32867624532295",
            "numberOfShares": "66.29717"
        },
        {
            "timestamp": "2023-01-05T18:01:44.028Z",
            "transactionSignature": "3kK68DRTnriVXwvkNenQQ27j7p9AyGnsf2RrrbDe9CJ2fwwNFQAX1C1WKuichHFpo6NZcVUL2MuhhxPEjLTGYPF1",
            "transactionName": "depositAndInvest",
            "strategy": "HWg7yB3C1BnmTKFMU3KGD7E96xx2rUhv4gxrwbZLXHBt",
            "tokenA": "SOL",
            "tokenAAmount": "0.35802676",
            "tokenAPrice": "13.08655",
            "tokenB": "BONK",
            "tokenBAmount": "2869729.18746",
            "tokenBPrice": "0.0000028564",
            "usdValue": "12.882429547138744",
            "numberOfShares": "10.079574"
        },
        {
            "timestamp": "2023-01-05T17:05:04.940Z",
            "transactionSignature": "3DzJuviUor15PYZovKJuD2gajysX34P54FCCzXe692ULUiDntZwALFzwA4owgJkhJqhGooA1DThE93EpN1edrZmC",
            "transactionName": "depositAndInvest",
            "strategy": "HWg7yB3C1BnmTKFMU3KGD7E96xx2rUhv4gxrwbZLXHBt",
            "tokenA": "SOL",
            "tokenAAmount": "0.16108004",
            "tokenAPrice": "13.18014",
            "tokenB": "BONK",
            "tokenBAmount": "4728878.00767",
            "tokenBPrice": "0.0000022961",
            "usdValue": "12.981034271816687",
            "numberOfShares": "12.127261"
        },
        {
            "timestamp": "2023-01-05T17:04:01.648Z",
            "transactionSignature": "2BasTEsBfwLucMt5UJWTkw41f4sesYZhcqjMzVQnZnJxTUKm5BwM4etHuFKLjVtQjsyeaGXLCmES6t4FsToGfFuk",
            "transactionName": "depositAndInvest",
            "strategy": "8XgX1EkSHC43mwdaUCZeXL2JVFz15JJynFrcxrQa3jss",
            "tokenA": "USDH",
            "tokenAAmount": "0.193515",
            "tokenAPrice": "0.9980399162",
            "tokenB": "BONK",
            "tokenBAmount": "342429.43023",
            "tokenBPrice": "0.0000022795",
            "usdValue": "0.973703580592728",
            "numberOfShares": "1.109471"
        },
        {
            "timestamp": "2023-01-05T17:01:05.584Z",
            "transactionSignature": "TKC6MzEtvRvaWSLk8R3hn2pE5qj1uFaonPKekbxyxCa2qosR7MjXEiuHmc68phgyYCLwodWHZcphMo11MtiGxj5",
            "transactionName": "depositAndInvest",
            "strategy": "8XgX1EkSHC43mwdaUCZeXL2JVFz15JJynFrcxrQa3jss",
            "tokenA": "USDH",
            "tokenAAmount": "1",
            "tokenAPrice": "0.9980145446",
            "tokenB": "BONK",
            "tokenBAmount": "1841068.38417",
            "tokenBPrice": "0.0000022704",
            "usdValue": "5.177976204019568",
            "numberOfShares": "5.919282"
        },
        {
            "timestamp": "2023-01-05T13:56:43.218Z",
            "transactionSignature": "YR3WvYfQFtsnYCqwbj9nZ8D3GknyqHjy6CT87i9tJ659dC2xe9X9tVarP36bGnYUM5hgAEGtEPJnKCDapk8KohX",
            "transactionName": "depositAndInvest",
            "strategy": "8XgX1EkSHC43mwdaUCZeXL2JVFz15JJynFrcxrQa3jss",
            "tokenA": "USDH",
            "tokenAAmount": "1",
            "tokenAPrice": "0.9979248774",
            "tokenB": "BONK",
            "tokenBAmount": "607043.94701",
            "tokenBPrice": "0.000002961",
            "usdValue": "2.79538200449661",
            "numberOfShares": "2.605278"
        },
        {
            "timestamp": "2023-01-05T09:23:04.301Z",
            "transactionSignature": "2DA7c9qLATV7qBbraSzy4UBybwExrqKHhEj9wMgFF83rAnUvw1oeV51Cw1qerDyBaNJ1ANBGZ7QWhg66dUR62vV5",
            "transactionName": "depositAndInvest",
            "strategy": "8XgX1EkSHC43mwdaUCZeXL2JVFz15JJynFrcxrQa3jss",
            "tokenA": "USDH",
            "tokenAAmount": "0.540482",
            "tokenAPrice": "0.9980125319",
            "tokenB": "BONK",
            "tokenBAmount": "108091.05922",
            "tokenBPrice": "0.0000041745",
            "usdValue": "0.9906339359802658",
            "numberOfShares": "0.717938"
        },
        {
            "timestamp": "2023-01-05T00:53:07.073Z",
            "transactionSignature": "5FmpDytoHx7aLH12ubzzrspse1PFV2URYd4zPHzZ7Pj2LfAsQ2i5sybnk13BLwZbHNQEfNG6BMfxsn29nQNtm9RA",
            "transactionName": "depositAndInvest",
            "strategy": "HWg7yB3C1BnmTKFMU3KGD7E96xx2rUhv4gxrwbZLXHBt",
            "tokenA": "SOL",
            "tokenAAmount": "0.179316875",
            "tokenAPrice": "13.21",
            "tokenB": "BONK",
            "tokenBAmount": "554845.76237",
            "tokenBPrice": "0.0000033614",
            "usdValue": "4.233834464380518",
            "numberOfShares": "3.372756"
        },
        {
            "timestamp": "2023-01-05T00:50:32.064Z",
            "transactionSignature": "5msg4zqTK3S7bpiVjZZs6xWqg9Hw8r4thRnyR2gk68WNL4SxEVpFTmw7RC8eW17Ubhf46sLy3FWM3Aeg5Ugjuufu",
            "transactionName": "depositAndInvest",
            "strategy": "HWg7yB3C1BnmTKFMU3KGD7E96xx2rUhv4gxrwbZLXHBt",
            "tokenA": "SOL",
            "tokenAAmount": "1",
            "tokenAPrice": "13.17",
            "tokenB": "BONK",
            "tokenBAmount": "3699754.86054",
            "tokenBPrice": "0.0000032746",
            "usdValue": "25.285217266324284",
            "numberOfShares": "20.406608"
        },
        {
            "timestamp": "2023-01-05T00:39:55.025Z",
            "transactionSignature": "42UrLcWiPUaxXVszQiitZDb7YbDV1ML1a1zjCjKP9XFHHMLewbZnV6op9jX7wazpvwuG6zG62aGVU9rUNSZxh64s",
            "transactionName": "depositAndInvest",
            "strategy": "HWg7yB3C1BnmTKFMU3KGD7E96xx2rUhv4gxrwbZLXHBt",
            "tokenA": "SOL",
            "tokenAAmount": "1.000005",
            "tokenAPrice": "13.34343025",
            "tokenB": "BONK",
            "tokenBAmount": "6141937.87297",
            "tokenBPrice": "0.0000031716",
            "usdValue": "32.823267125062902",
            "numberOfShares": "26.698485"
        },
        {
            "timestamp": "2023-01-05T00:39:30.350Z",
            "transactionSignature": "3dCrFS6dckRqmgUYXT61t3uELc9mwc3dzPUnaA48uga9Mn6uzLfDnHCh9rWJPpXZ8CsofCcEU8KcLdmGPrmpBwYj",
            "transactionName": "depositAndInvest",
            "strategy": "HWg7yB3C1BnmTKFMU3KGD7E96xx2rUhv4gxrwbZLXHBt",
            "tokenA": "SOL",
            "tokenAAmount": "1",
            "tokenAPrice": "13.34343025",
            "tokenB": "BONK",
            "tokenBAmount": "1",
            "tokenBPrice": "0.0000031716",
            "usdValue": "13.3434334216",
            "numberOfShares": "0.000004"
        },
        {
            "timestamp": "2023-01-04T21:55:52.127Z",
            "transactionSignature": "5qEswBDBdkcsq9JeqdRp96YUGYksvt3TP3EAfbEwm5qGMXmoW9cQXziAh1GbNYk3iBcCDM1s7Ao5cfu4KxmSFqwA",
            "transactionName": "depositAndInvest",
            "strategy": "98kNMp1aqWoYAaUU8m5REBAYVwhFb4aX9yoSpgq8kUFu",
            "tokenA": "USDC",
            "tokenAAmount": "0.152401",
            "tokenAPrice": "1",
            "tokenB": "USDT",
            "tokenBAmount": "0.847422",
            "tokenBPrice": "0.9997185",
            "usdValue": "0.999584450707",
            "numberOfShares": "0.745073"
        },
        {
            "timestamp": "2023-01-04T21:18:17.982Z",
            "transactionSignature": "3HqCueKoHPGk3uMuCJTNEw4g76ozwPR1aqsPPyKUT82BbR1gX6ApqEoq9Qh7HazEBGnrCAaVKrVJmagrk5GvbR69",
            "transactionName": "depositAndInvest",
            "strategy": "HWg7yB3C1BnmTKFMU3KGD7E96xx2rUhv4gxrwbZLXHBt",
            "tokenA": "SOL",
            "tokenAAmount": "0.09537866",
            "tokenAPrice": "13.025",
            "tokenB": "BONK",
            "tokenBAmount": "51718.75246",
            "tokenBPrice": "0.0000024596",
            "usdValue": "1.369514490050616",
            "numberOfShares": "1.144917"
        },
        {
            "timestamp": "2023-01-04T21:13:42.819Z",
            "transactionSignature": "2hUxqS37Kx8yTey7VPDko2Ytn8WNxHj8tojMgPoNhG7wQo2NKhgxH8WDuxjJPrgqds4W9Z88wr9k78bYCMBYHUTf",
            "transactionName": "depositAndInvest",
            "strategy": "ByXB4xCxVhmUEmQj3Ut7byZ1Hbva1zhKjaVcv3jBMN7E",
            "tokenA": "SOL",
            "tokenAAmount": "0.636271159",
            "tokenAPrice": "13.0263",
            "tokenB": "STSOL",
            "tokenBAmount": "0.337429654",
            "tokenBPrice": "14.0337464001",
            "usdValue": "13.0236611905911885654",
            "numberOfShares": "0.042788929"
        },
        {
            "timestamp": "2023-01-04T21:05:22.636Z",
            "transactionSignature": "2iXseS1QpK58UJbHnmfPrWxZmcXjvgQV1FpVbRqo1b3mPEG6QbZBaqWJUwLuG2ct3tnafHLLC8ThdRKHg5kRKyRA",
            "transactionName": "depositAndInvest",
            "strategy": "HWg7yB3C1BnmTKFMU3KGD7E96xx2rUhv4gxrwbZLXHBt",
            "tokenA": "SOL",
            "tokenAAmount": "1",
            "tokenAPrice": "13.06354",
            "tokenB": "BONK",
            "tokenBAmount": "1215855.35737",
            "tokenBPrice": "0.0000023436",
            "usdValue": "15.913018615532332",
            "numberOfShares": "13.377477"
        },
        {
            "timestamp": "2023-01-04T17:53:08.618Z",
            "transactionSignature": "4TQV4ZGZMQ6orAUis6iknQSvbDUFoRXeKgNmfwSC4qBNDtcbBdid6w6i3KWsAGoxjyvhcMJiddQ9SH3kN1sv2RsP",
            "transactionName": "depositAndInvest",
            "strategy": "Cfuy5T6osdazUeLego5LFycBQebm9PP3H7VNdCndXXEN",
            "tokenA": "USDH",
            "tokenAAmount": "5.659653",
            "tokenAPrice": "0.9980770889",
            "tokenB": "USDC",
            "tokenBAmount": "4.318876",
            "tokenBPrice": "0.9999525",
            "usdValue": "9.9674408438141517",
            "numberOfShares": "9.890922"
        },
        {
            "timestamp": "2023-01-03T21:34:35.185Z",
            "transactionSignature": "hKPq3VUynLYgcTU45odaRx3Cno67ZF6qkPyqoKgd9wduXU6MVzzQfkvYKwzJMZog7vGXcH8jhe33qzbeqVXXGUw",
            "transactionName": "depositAndInvest",
            "strategy": "HWg7yB3C1BnmTKFMU3KGD7E96xx2rUhv4gxrwbZLXHBt",
            "tokenA": "SOL",
            "tokenAAmount": "0.351976542",
            "tokenAPrice": "13.10210673",
            "tokenB": "BONK",
            "tokenBAmount": "1501252.48945",
            "tokenBPrice": "0.0000010992",
            "usdValue": "6.26181095614376766",
            "numberOfShares": "5.478513"
        },
        {
            "timestamp": "2023-01-03T21:32:18.288Z",
            "transactionSignature": "4HrGWQb6Ek1saqEqu7Lo5vtQU7FNzkdkacskESuiM9uUzqXmjLBiNJQJyvi4uJCK4ZcBHM3aPhW7sViGN2JRq8xz",
            "transactionName": "depositAndInvest",
            "strategy": "HWg7yB3C1BnmTKFMU3KGD7E96xx2rUhv4gxrwbZLXHBt",
            "tokenA": "SOL",
            "tokenAAmount": "0.674983383",
            "tokenAPrice": "13.12083781",
            "tokenB": "BONK",
            "tokenBAmount": "3132759.42848",
            "tokenBPrice": "0.0000011065",
            "usdValue": "12.32274580040123123",
            "numberOfShares": "10.751346"
        },
        {
            "timestamp": "2023-01-03T21:29:25.563Z",
            "transactionSignature": "2Cp6sEgebZPrhho2p7vn3R1EZaYvNJ3rRSsEkpz9c1ZBSXQMqtMQxusfZX6ht9CFPPqqL5r985aaqE6UP1h9WJtB",
            "transactionName": "depositAndInvest",
            "strategy": "HWg7yB3C1BnmTKFMU3KGD7E96xx2rUhv4gxrwbZLXHBt",
            "tokenA": "SOL",
            "tokenAAmount": "0.71629186",
            "tokenAPrice": "13.10045896",
            "tokenB": "BONK",
            "tokenBAmount": "3055825.13306",
            "tokenBPrice": "0.0000011195",
            "usdValue": "12.8047483517727356",
            "numberOfShares": "11.153154"
        },
        {
            "timestamp": "2023-01-03T21:17:52.418Z",
            "transactionSignature": "4AcuccPfpzZdPZYNTCtKbjzxCGDXBLG7ZmGWnsqdoqg8tQFA2yFuxQWsP9ckKw4decP2JLdSzgmymaN4a7P7frEY",
            "transactionName": "depositAndInvest",
            "strategy": "HWg7yB3C1BnmTKFMU3KGD7E96xx2rUhv4gxrwbZLXHBt",
            "tokenA": "SOL",
            "tokenAAmount": "0.763420284",
            "tokenAPrice": "13.10145594",
            "tokenB": "BONK",
            "tokenBAmount": "2735551.26002",
            "tokenBPrice": "0.0000011288",
            "usdValue": "13.08980747683886296",
            "numberOfShares": "11.382516"
        },
        {
            "timestamp": "2023-01-03T21:10:17.307Z",
            "transactionSignature": "3FQkUS8TmP6fBvmSREpuSSLrkSaB4AbSURg6TSPMqovyoNPtZNVuT853PXpD2i1fS7s3xweAeYmYHUEjXa28k7cz",
            "transactionName": "depositAndInvest",
            "strategy": "HWg7yB3C1BnmTKFMU3KGD7E96xx2rUhv4gxrwbZLXHBt",
            "tokenA": "SOL",
            "tokenAAmount": "1.000005",
            "tokenAPrice": "13.01200265",
            "tokenB": "BONK",
            "tokenBAmount": "2544050.17918",
            "tokenBPrice": "0.00000116",
            "usdValue": "15.96316591786205",
            "numberOfShares": "13.878382"
        },
        {
            "timestamp": "2023-01-03T21:08:46.296Z",
            "transactionSignature": "pYNnJxJt4MQFq5RE4LcTRErv2vi2KUynEEFYqsA5fZZ3fpTVUcMpWzMwbB3Czh2G6HvNUA3cHNdjE6YvrSCyFAv",
            "transactionName": "depositAndInvest",
            "strategy": "HWg7yB3C1BnmTKFMU3KGD7E96xx2rUhv4gxrwbZLXHBt",
            "tokenA": "SOL",
            "tokenAAmount": "1.000005",
            "tokenAPrice": "13.02964095",
            "tokenB": "BONK",
            "tokenBAmount": "2433339.34823",
            "tokenBPrice": "0.0000011648",
            "usdValue": "15.864059771023054",
            "numberOfShares": "13.767199"
        },
        {
            "timestamp": "2022-12-19T17:27:55.494Z",
            "transactionSignature": "5oSgVmRxr22Ts6Wd44pHJB5XBbktpYs5jjzYpCpjW1y9oC2GeTa1ebzjCU48wBpTmQUoJYbm97rTUEBzHeYm5XbZ",
            "transactionName": "depositAndInvest",
            "strategy": "98kNMp1aqWoYAaUU8m5REBAYVwhFb4aX9yoSpgq8kUFu",
            "tokenA": "USDC",
            "tokenAAmount": "0.5",
            "tokenAPrice": "0.99995001",
            "tokenB": "USDT",
            "tokenBAmount": "0.88215",
            "tokenBPrice": "0.99997",
            "usdValue": "1.3820985405",
            "numberOfShares": "1.030552"
        },
        {
            "timestamp": "2022-12-17T21:57:03.132Z",
            "transactionSignature": "Ed4YESigpxLouqXVbmGCfw7T7HAQHNC9dNoVqecccfu6JTtq8bmB4GfqghnEG3XQu988FVGKkqDreW4Y6ESwPzd",
            "transactionName": "depositAndInvest",
            "strategy": "AL6Yd51aSY3S9wpuypJYKtEf65xBSXKpfUCxN54CaWeE",
            "tokenA": "SOL",
            "tokenAAmount": "0.074925693",
            "tokenAPrice": "12.295",
            "tokenB": "USDC",
            "tokenBAmount": "0.134446",
            "tokenBPrice": "0.99996202",
            "usdValue": "1.05565228917592",
            "numberOfShares": "1.070001"
        },
        {
            "timestamp": "2023-01-05T17:48:50.643Z",
            "transactionSignature": "3cNSZgBrojvQQbTP2yMMDTJQzo7dZXdtQFDeoMbbU2A3rY7XY7dWFf7BgjTKhKcgQrKFPw9r3ZrGNUDGMXxCZw9D",
            "transactionName": "withdraw",
            "strategy": "HWg7yB3C1BnmTKFMU3KGD7E96xx2rUhv4gxrwbZLXHBt",
            "tokenA": "SOL",
            "tokenAAmount": "3.542053163",
            "tokenAPrice": "13.085",
            "tokenB": "BONK",
            "tokenBAmount": "48021209.75745",
            "tokenBPrice": "0.0000025358",
            "usdValue": "168.11994934079671",
            "numberOfShares": "143.538618"
        },
        {
            "timestamp": "2023-01-05T17:01:22.518Z",
            "transactionSignature": "5HBmFPpFQj3umaJgWhsC3LR5dD7jGHdq57fQwnC2mdVwFk6JdKUZ8Aw9gtCyqQLPXhAXkUecVSi9vYW5oumcbuWk",
            "transactionName": "withdraw",
            "strategy": "8XgX1EkSHC43mwdaUCZeXL2JVFz15JJynFrcxrQa3jss",
            "tokenA": "USDH",
            "tokenAAmount": "0.195435",
            "tokenAPrice": "0.9980145446",
            "tokenB": "BONK",
            "tokenBAmount": "354038.1496",
            "tokenBPrice": "0.0000022704",
            "usdValue": "0.998855187375741",
            "numberOfShares": "1.143631"
        },
        {
            "timestamp": "2023-01-05T09:25:53.571Z",
            "transactionSignature": "NRwMhUVmvJNiqnEgy7bYqZbQ2fsKB6r3p6zznDK3NfQBwLRZhtSxysj9eamL9XujZZEM6gFoke4VDBAiy14Lemn",
            "transactionName": "withdraw",
            "strategy": "8XgX1EkSHC43mwdaUCZeXL2JVFz15JJynFrcxrQa3jss",
            "tokenA": "USDH",
            "tokenAAmount": "0.053017",
            "tokenAPrice": "0.9980125345",
            "tokenB": "BONK",
            "tokenBAmount": "11063.79419",
            "tokenBPrice": "0.0000042129",
            "usdValue": "0.0995222890846375",
            "numberOfShares": "0.071935"
        },
        {
            "timestamp": "2022-12-19T17:28:37.766Z",
            "transactionSignature": "65nGWhvh4deZNYJzvzYEbUy9HzqF4UWCYroX94TB9hfhSYNhUEgErtn5GXCeBgycZpdPszRLa1wh3qcbGbE7DC5r",
            "transactionName": "withdraw",
            "strategy": "98kNMp1aqWoYAaUU8m5REBAYVwhFb4aX9yoSpgq8kUFu",
            "tokenA": "USDC",
            "tokenAAmount": "0.180862",
            "tokenAPrice": "0.99995001",
            "tokenB": "USDT",
            "tokenBAmount": "0.319096",
            "tokenBPrice": "0.99997611",
            "usdValue": "0.49994133550518",
            "numberOfShares": "0.372779"
        }
    ],
    "lastUpdatedOn": "2023-01-25T14:18:27.211Z"
}
```
